File: Python/054_Spiral_Matrix.py
"""
Given an m x n matrix, return all elements of the matrix in spiral order.
"""
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        res = []
        m, n = len(matrix), len(matrix[0])
        i, j = 0, 0
        while True:
            while j < n - 1 and matrix[i][j + 1] != 200:
                res.append(matrix[i][j])
                matrix[i][j] = 200
                j += 1
            while i < m - 1 and matrix[i + 1][j] != 200:
                res.append(matrix[i][j])
                matrix[i][j] = 200
                i += 1
            while j > 0 and matrix[i][j - 1] != 200:
                res.append(matrix[i][j])
                matrix[i][j] = 200
                j -= 1
            while i > 0 and matrix[i - 1][j] != 200:
                res.append(matrix[i][j])
                matrix[i][j] = 200
                i -= 1     
            if len(res) == m * n - 1:
                break
        res.append(matrix[i][j])
        return res               

File: Java/054_Spiral_Matrix.java
class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        int row = matrix.length, col = matrix[0].length;
        int [] res = new int[row * col];
        int i = 0, j = 0, cur = 0;
        while (true) {
            while (j < col - 1 && matrix[i][j + 1] != 101) {
                res[cur++] = matrix[i][j];
                matrix[i][j++] = 101;
            }
            while (i < row - 1 && matrix[i + 1][j] != 101) {
                res[cur++] = matrix[i][j];
                matrix[i++][j] = 101;
            }
            while (j > 0 && matrix[i][j - 1] != 101) {
                res[cur++] = matrix[i][j];
                matrix[i][j--] = 101;
            }
            while (i > 0 && matrix[i - 1][j] != 101) {
                res[cur++] = matrix[i][j];
                matrix[i--][j] = 101;
            }
            if (cur == row * col - 1) {
                break;
            }
        }
        res[cur] = matrix[i][j];
        return Arrays.stream(res).boxed().collect(Collectors.toList());
    }
}